include/execution.hpp

```diff
@@ -117,56 +119,125 @@ namespace std::execution {
+          template <template <class...> class _Variant>
+            using error_types = typename _Sender::template error_types<_Variant>;
+          static constexpr bool sends_done = _Sender::sends_done;
+        };
 
-  template <class _Sender>
-  auto __sender_traits_base_fn() {
-    if constexpr (__has_sender_types<_Sender>) {
-      return __typed_sender<_Sender>{};
-    } else if constexpr (derived_from<_Sender, sender_base>) {
-      return sender_base{};
-    } else if constexpr (__awaitable<_Sender>) {
-      if constexpr (is_void_v<__await_result_t<_Sender>>) {
-        return __sender_of<false>{};
-      } else {
-        return __sender_of<false, __await_result_t<_Sender>>{};
-      }
-    } else {
-      struct __no_sender_traits{
-        using __this_is_not_a_sender = _Sender;
+      struct get_sender_traits_t {
+        template <class _Sender, class _Receiver>
+        constexpr auto operator()(_Sender&&, _Receiver&&) const noexcept {
+          static_assert(sizeof(_Sender), "Incomplete type used with get_sender_traits");
+          static_assert(sizeof(_Receiver), "Incomplete type used with get_sender_traits");
+          if constexpr (tag_invocable<get_sender_traits_t, _Sender, _Receiver>) {
+            if constexpr (is_void_v<tag_invoke_result_t<get_sender_traits_t, _Sender, _Receiver>>) {
+              return sender_base{};
+            } else {
+              return tag_invoke_result_t<get_sender_traits_t, _Sender, _Receiver>{};
+            }
+          } else if constexpr (__has_sender_types<remove_cvref_t<_Sender>>) {
+            return __typed_sender<remove_cvref_t<_Sender>>{};
+          } else if constexpr (derived_from<remove_cvref_t<_Sender>, sender_base>) {
+            return sender_base{};
+          } else if constexpr (__awaitable<_Sender>) {
+            using _Result = __await_result_t<_Sender>;
+            if constexpr (is_void_v<_Result>) {
+              return receiver_signatures<set_value_t(), set_error_t(exception_ptr)>{};
+            } else {
+              return receiver_signatures<set_value_t(_Result), set_error_t(exception_ptr)>{};
+            }
+          } else {
+            struct __no_sender_traits{
+              using __this_is_not_a_sender = _Sender;
+            };
+            return __no_sender_traits{};
+          }
+        }
+
+        template <class _Sender>
+        constexpr auto operator()(_Sender&& __sndr) const noexcept {
+          return (*this)(__sndr, __default_context{});
+        }
       };
-      return __no_sender_traits{};
-    }
-  }
+    } // namespace __impl
+    using __impl::__default_context;
+
+    using __impl::get_sender_traits_t;
+    inline constexpr get_sender_traits_t get_sender_traits {};
+  } // namespace __sender_traits
 
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.traits]
-  template <class _Sender>
+  template <class _Sender, class _Receiver = __default_context>
```

Comment:
Currently sender traits serves two purposes: to distinguish typed senders from senders (and with what matedata), and to distinguish senders from non-senders. The "default context" is for the latter. If we want a "`sender`" concept that we can use when constraining the algorithms, then we need some way of asking a type: are you a sender-ish thing?

I'll admit this isn't a very EoP concept. It has no semantics whatsoever, and apart from move constructibility, it has no syntactic requirements except those we impose. So there's nothing wrong _per se_ with dropping `sender` as a unary concept. We could just leave the sender argument to the algorithms unconstrained.

Reply:
There could be things that are senders but that require a particular kind of receiver to be able to act as a sender.

Reply:
True. I'm not arguing that point, though. In the sender/receiver ontology, such a thing is still has sender-ness: it is a unit of work, possibly async, that has not been scheduled for execution yet. It is certainly an error if you pass it where, say, a scheduler or a receiver is expected. And if I pass a scheduler or a receiver where a sender is expected, I would also like that to be an error, I think. The question is: how do we make that an error?

In the past, we've used `sender_traits` to say that something is a sender in the async ontology. My question is: do we still want to do that?

Reply:
I think that we need some way to be able to say that a type is a sender absent a particular receiver that it will be connected to. I don't know if `get_sender_traits` should be that thing, however.

I guess we _could_ have `get_sender_traits()` have two flavours - one that takes only a sender and one that takes a sender and a receiver. The one taking only a sender could be used to determine whether a type is a sender (e.g. if the call expression is well-formed), and could possibly list the completion signals if they are independent of the receiver (e.g. by returning a non-void type).

However, such an approach would complicate the approach of calculating the completion signals for sender adapters.
e.g. a sender adapter when reporting its own `get_sender_traits(sender)` result would need to check whether the wrapped sender reported actual traits or not from `get_sender_traits(inner_sender)` and then conditionally return a sender-traits type that computed the right results.

Given that most sender's traits types will likely be dependent on the receiver (if only for reporting a sends_done based on the receiver's stop token) I'm not sure it makes sense to still support the ability to query the traits for the sender independently of the receiver.

Maybe there should just be some independent way of querying whether a type is a sender rather than combining this with `get_sender_traits(s, r)` query?

e.g. maybe something like:
```c++
struct my_sender {
  friend std::true_type tag_invoke(std::execution::is_sender, const my_sender&);
  // ...
};
```

Reply:
I've taken a stab at a design that uses `get_sender_traits(snd)` to distinguish between senders and non-senders. I have preserved the ability to ask a sender if it is typed independent of the receiver. I haven't found it to complicate the computation of the completion signals. The trick is for `get_sender_traits` to default the context parameter to a special `no_context` type. `no_context` is special because all ~~receiver~~ contextual queries are ill-formed when passed `no_context`.

In addition, `typed_sender`s have a semantic requirement that _if_ they are typed both with and without a context, their completion signals must be the same. When compiling in debug mode, this is enforced statically.